File: app/src/main/java/ua/com/foxminded/locationtrackera/ui/auth/registration/RegistrationViewModel.java
package ua.com.foxminded.locationtrackera.ui.auth.registration;

import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.google.firebase.auth.FirebaseAuthUserCollisionException;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;
import io.reactivex.rxjava3.subjects.PublishSubject;
import ua.com.foxminded.locationtrackera.R;
import ua.com.foxminded.locationtrackera.model.auth.AuthNetwork;
import ua.com.foxminded.locationtrackera.mvi.MviViewModel;
import ua.com.foxminded.locationtrackera.ui.auth.AuthErrorConstants;
import ua.com.foxminded.locationtrackera.ui.auth.Credentials;
import ua.com.foxminded.locationtrackera.ui.auth.registration.state.RegistrationScreenEffect;
import ua.com.foxminded.locationtrackera.ui.auth.registration.state.RegistrationScreenState;
import ua.com.foxminded.locationtrackera.util.Result;

public class RegistrationViewModel extends MviViewModel<RegistrationScreenState, RegistrationScreenEffect>
        implements RegistrationContract.ViewModel {

    private final PublishSubject<Credentials> credsSupplier = PublishSubject.create();
    private final AuthNetwork authNetwork;

    public RegistrationViewModel(AuthNetwork authNetwork) {
        this.authNetwork = authNetwork;
    }

    @Override
    public void onStateChanged(@NonNull LifecycleOwner source, @NonNull Lifecycle.Event event) {
        super.onStateChanged(source, event);
        if (event == Lifecycle.Event.ON_CREATE) {
            setUpRegistrationChain();
        }
    }

    @SuppressWarnings("unchecked")
    private void setUpRegistrationChain() {
        addTillDestroy(
                credsSupplier.observeOn(Schedulers.io())
                        .flatMapSingle(creds -> {
                            if (creds.isUsernameValid() && creds.isEmailValid() && creds.isRegistrationPasswordValid()) {
                                postState(new RegistrationScreenState.RegistrationProgress(true));
                                return authNetwork.firebaseRegister(creds.username, creds.email, creds.password);
                            } else {
                                postState(getErrorState(creds));
                                return Single.just(new Result.Error<>(new Throwable(AuthErrorConstants.INVALID_USERNAME_EMAIL_PASSWORD)));
                            }
                        }).observeOn(AndroidSchedulers.mainThread())
                        .subscribe(result -> {
                            if (result.isSuccessful()) {
                                setState(new RegistrationScreenState.RegistrationProgress(false));
                                setAction(new RegistrationScreenEffect.RegistrationSuccessful());
                            } else {
                                if (!result.toString().contains(AuthErrorConstants.INVALID_USERNAME_EMAIL_PASSWORD)) {
                                    setState(new RegistrationScreenState.RegistrationProgress(false));
                                    if (((Result.Error<Void>) result).getError() instanceof FirebaseAuthUserCollisionException) {
                                        setAction(new RegistrationScreenEffect.RegistrationFailed(R.string.user_already_exists));
                                    } else {
                                        setAction(new RegistrationScreenEffect.RegistrationFailed(R.string.registration_failed));
                                    }
                                }
                            }
                        }, error -> {
                            error.printStackTrace();
                            setState(new RegistrationScreenState.RegistrationProgress(false));
                            setAction(new RegistrationScreenEffect.RegistrationFailed(R.string.registration_failed));
                        })
        );
    }

    private RegistrationScreenState getErrorState(Credentials creds) {
        int usernameError, emailError, passwordError;
        if (creds.isUsernameValid()) {
            usernameError = 0;
        } else {
            usernameError = R.string.empty_field;
        }
        if (creds.isEmailValid()) {
            emailError = 0;
        } else {
            emailError = R.string.invalid_email;
        }
        if (creds.isRegistrationPasswordValid()) {
            passwordError = 0;
        } else {
            passwordError = R.string.invalid_password;
        }
        return new RegistrationScreenState.RegistrationError(usernameError, emailError, passwordError);
    }

    public void registerUser(String username, String email, String password) {
        credsSupplier.onNext(new Credentials(username, email, password));
    }

    public void registrationDataChanged(String username, String email, String password) {
        setState(getErrorState(new Credentials(username, email, password)));
    }
}

File: app/src/main/java/ua/com/foxminded/locationtrackera/ui/auth/registration/RegistrationViewModel.kt
package ua.com.foxminded.locationtrackera.ui.auth.registration

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import ua.com.foxminded.locationtrackera.R
import ua.com.foxminded.locationtrackera.models.auth.AuthNetwork
import ua.com.foxminded.locationtrackera.models.util.Result
import ua.com.foxminded.locationtrackera.mvi.MviViewModel
import ua.com.foxminded.locationtrackera.ui.auth.AuthErrorConstants
import ua.com.foxminded.locationtrackera.ui.auth.Credentials
import ua.com.foxminded.locationtrackera.ui.auth.registration.state.RegistrationScreenEffect
import ua.com.foxminded.locationtrackera.ui.auth.registration.state.RegistrationScreenState

class RegistrationViewModel(private val authNetwork: AuthNetwork) : MviViewModel<
        RegistrationScreenState,
        RegistrationScreenEffect>(),
    RegistrationContract.ViewModel {

    private val registerFlow = MutableSharedFlow<Credentials>(extraBufferCapacity = 1)
    private var launch: Job? = null
    private var errorCreds: Credentials? = null

    override fun onStateChanged(event: Lifecycle.Event) {
        super.onStateChanged(event)
        if (event == Lifecycle.Event.ON_CREATE) {
            launch = viewModelScope.launch {
                registerFlow
                    .onEach { setState(RegistrationScreenState.RegistrationProgress(true)) }
                    .flowOn(Dispatchers.Main)
                    .map { creds ->
                        if (creds.isUsernameValid() && creds.isEmailValid() && creds.isRegistrationPasswordValid()) {
                            return@map authNetwork.firebaseRegister(
                                creds.username!!, creds.email!!, creds.password!!
                            )
                        } else {
                            errorCreds = creds
                            return@map Result.Error(
                                Throwable(AuthErrorConstants.INVALID_USERNAME_EMAIL_PASSWORD)
                            )
                        }
                    }
                    .flowOn(Dispatchers.IO)
                    .catch { error ->
                        error.printStackTrace()
                        setState(RegistrationScreenState.RegistrationProgress(false))
                        setAction(RegistrationScreenEffect.RegistrationFailed(R.string.registration_failed))
                    }
                    .collect { result ->
                        setState(RegistrationScreenState.RegistrationProgress(false))
                        if (result.isSuccessful) {
                            setAction(RegistrationScreenEffect.RegistrationSuccessful())
                        } else {
                            if (result.toString()
                                    .contains(AuthErrorConstants.INVALID_USERNAME_EMAIL_PASSWORD)
                            ) {
                                setState(getErrorState(errorCreds))
                            } else {
                                if ((result as Result.Error<Void?>).error is FirebaseAuthUserCollisionException) {
                                    setAction(RegistrationScreenEffect.RegistrationFailed(R.string.user_already_exists))
                                } else {
                                    setAction(RegistrationScreenEffect.RegistrationFailed(R.string.registration_failed))
                                }
                            }
                        }
                    }
            }
        }
    }

    private fun getErrorState(creds: Credentials?): RegistrationScreenState {
        var usernameError = 0
        var emailError = 0
        var passwordError = 0
        creds?.let {
            usernameError = if (creds.isUsernameValid()) 0 else R.string.empty_field
            emailError = if (creds.isEmailValid()) 0 else R.string.invalid_email
            passwordError = if (creds.isRegistrationPasswordValid()) 0 else R.string.invalid_password
        }
        return RegistrationScreenState.RegistrationError(usernameError, emailError, passwordError)
    }

    override fun registerUser(username: String?, email: String?, password: String?) {
        registerFlow.tryEmit(Credentials(username, email, password))
    }

    override fun registrationDataChanged(username: String?, email: String?, password: String?) {
        setState(getErrorState(Credentials(username, email, password)))
    }

    override fun onCleared() {
        launch = null
        super.onCleared()
    }
}
